internal/app/server/server.go

```diff
@@ -0,0 +1,126 @@
+
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"connectrpc.com/conformance/internal/gen/proto/connect/connectrpc/conformance/v1alpha1/conformancev1alpha1connect"
+	"github.com/rs/cors"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
+)
+
+func Run(ctx context.Context, args []string, in io.ReadCloser, out, err io.WriteCloser) error {
+	h1Port := args[0]
+	h2Port := args[1]
+
+	mux := http.NewServeMux()
+	mux.Handle(conformancev1alpha1connect.NewConformanceServiceHandler(
+		NewConformanceServiceHandler(),
+	))
+	corsHandler := cors.New(cors.Options{
+		AllowedMethods: []string{
+			http.MethodHead,
+			http.MethodGet,
+			http.MethodPost,
+			http.MethodPut,
+			http.MethodPatch,
+			http.MethodDelete,
+		},
+		// Mirror the `Origin` header value in the `Access-Control-Allow-Origin`
+		// preflight response header.
+		// This is equivalent to `Access-Control-Allow-Origin: *`, but allows
+		// for requests with credentials.
+		// Note that this effectively disables CORS and is not safe for use in
+		// production environments.
+		AllowOriginFunc: func(origin string) bool {
+			return true
+		},
+		// Note that rs/cors does not return `Access-Control-Allow-Headers: *`
+		// in response to preflight requests with the following configuration.
+		// It simply mirrors all headers listed in the `Access-Control-Request-Headers`
+		// preflight request header.
+		AllowedHeaders: []string{"*"},
+		// We explicitly set the exposed header names instead of using the wildcard *,
+		// because in requests with credentials, it is treated as the literal header
+		// name "*" without special semantics.
+		ExposedHeaders: []string{
+			"Grpc-Status", "Grpc-Message", "Grpc-Status-Details-Bin", "X-Grpc-Test-Echo-Initial",
+			"Trailer-X-Grpc-Test-Echo-Trailing-Bin", "Request-Protocol", "Get-Request"},
+	}).Handler(mux)
+
+	// Create servers
+	h1Server := newH1Server(h1Port, corsHandler)
+	h2Server := newH2Server(h2Port, mux)
+	done := make(chan os.Signal, 1)
+	errs := make(chan error, 2)
+	signal.Notify(done, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		err := h1Server.ListenAndServe()
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			errs <- err
+		}
+	}()
+	go func() {
+		err := h2Server.ListenAndServe()
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			errs <- err
+		}
+	}()
+
+	fmt.Printf("HTTP/1.1 server listening on port %s\nHTTP/2 server listening on port %s", h1Port, h2Port)
+
+	select {
+	case err := <-errs:
+		return err
+	case <-done:
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	if err := h1Server.Shutdown(ctx); err != nil {
+		return err
+	}
+	if err := h2Server.Shutdown(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func newH1Server(h1Port string, handler http.Handler) *http.Server {
+	h1Server := &http.Server{
+		Addr:    ":" + h1Port,
```

Comment:
I think this is equivalent to `"0.0.0.0:" + port`, which exposes the server on all network interfaces. But it is likely adequate to instead use `"127.0.0.1:" + port` to only expose the server on the loopback interface. It should probably at least _default_ to the loopback interface, and we could make it a parameter to bind to other interfaces if needed (unclear if it will be though).